coderd/workspaceagentsrpc.go

```diff
@@ -184,73 +168,159 @@ func ensureLatestBuild(ctx context.Context, db database.Store, logger slog.Logge
 			Message: "Agent trying to connect from non-latest build.",
 			Detail:  err.Error(),
 		})
-		return nil, database.WorkspaceBuild{}, false
+		return database.WorkspaceBuild{}, false
 	}
 
-	return ensureLatestBuild, build, true
+	return build, true
 }
 
-func (api *API) agentConnectionUpdate(ctx context.Context, workspaceAgent database.WorkspaceAgent, workspaceID uuid.UUID, conn *websocket.Conn) (func(closeCtx context.Context, ensureLatestBuildFn func() error), bool) {
-	// We use a custom heartbeat routine here instead of `httpapi.Heartbeat`
-	// because we want to log the agent's last ping time.
-	var lastPing atomic.Pointer[time.Time]
-	lastPing.Store(ptr.Ref(time.Now())) // Since the agent initiated the request, assume it's alive.
-
-	go pprof.Do(ctx, pprof.Labels("agent", workspaceAgent.ID.String()), func(ctx context.Context) {
-		// TODO(mafredri): Is this too frequent? Use separate ping disconnect timeout?
-		t := time.NewTicker(api.AgentConnectionUpdateFrequency)
-		defer t.Stop()
-
-		for {
-			select {
-			case <-t.C:
-			case <-ctx.Done():
-				return
-			}
+func checkBuildIsLatest(ctx context.Context, db database.Store, build database.WorkspaceBuild) error {
+	latestBuild, err := db.GetLatestWorkspaceBuildByWorkspaceID(ctx, build.WorkspaceID)
+	if err != nil {
+		return err
+	}
+	if build.ID != latestBuild.ID {
+		return xerrors.New("build is outdated")
+	}
+	return nil
+}
 
-			// We don't need a context that times out here because the ping will
-			// eventually go through. If the context times out, then other
-			// websocket read operations will receive an error, obfuscating the
-			// actual problem.
-			err := conn.Ping(ctx)
-			if err != nil {
-				return
-			}
-			lastPing.Store(ptr.Ref(time.Now()))
+func (api *API) startAgentWebsocketMonitor(ctx context.Context,
```

Comment:
Since we now have a close method, I don't think we need to take a context here anymore, wdyt? For `start` we could simply create the context/cancel func based on api context here and pass it along.

PS. Thanks for the naming change, this now feels very obvious what it does when reading the code!

Reply:
No, the API context is used when sending the final updated that the agent has been disconnected.  

The context we accept here, and want for `start()`, is tied to the connection.